### ISSUE TITLE
Consider making IDE0044 (Make field readonly) precondition more strict/accurate

### ISSUE THREAD

**Issue by GeeLaw** [dotnet/roslyn#85223]
**Brief description:**

Currently, IDE0044 is emitted whenever there's no assignment to a field outside of the constructor. However, adding `readonly` can alter semantics as methods might mutate a `struct`.

**Languages applicable:** Applicable to C# and VB.

**Code example that the analyzer should report:**

```csharp
public interface IFoo { void Init(); void DoWork(); }

public class C<T> where T : struct, IFoo
{
  // IDE0044: Make field readonly
  // Prior to this issue: IDE0044 is reported.
  // Suggestion: Do not report IDE0044 here.
  private T foo;

  // Two initialization modes available.
  public C() { foo.Init(); }
  public C(T f) { foo = f; }

  public void DoTwice()
  {
    foo.DoWork();
    foo.DoWork();
  }
}
```

A real-world scenario is when `T` above represents a hash function that's resampleable, and `Init` samples one from the family.

**Additional information:**

The condition of reporting IDE0044 roughly has some similarities with that of simplifying identity conversions. (See #85201 ) I propose the following condition to be used for non-byref fields.

- If the field is assigned to outside of the constructors, **do not report** IDE0044 (this rule already exists). Otherwise, consider the rules below until a decision is first made.
- If the field is of reference type (a type-parameter type constrained to `class` or `class?` or a base class that's not `Enum`, or a reference type constructed from 0 or more type parameters), then IDE0044 **should be reported**.
- If the field is of value type that is known to be `readonly` (in this case, it must be a value type constructed from 0 or more type parameters that is not itself a type parameter), then IDE0044 **should be reported**.
- If the field type is pointer, then IDE0044 **should be reported**.
- Otherwise, **do not report** IDE0044.

Moreover, currently IDE0044 is not reported for byref fields. The proposed rule for byref fields is as follows.

- If the field is ref-assigned-to outside of the constructors, **do not report** IDE0044.
- Otherwise, IDE0044 **should be reported**.

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

CC @jnm2

**Comment by MihuBot** [comment]
I'm a bot. Here are possible related and/or duplicate issues (I may be wrong):
- https://github.com/dotnet/roslyn/issues/49290
- https://github.com/dotnet/roslyn/issues/30240
- https://github.com/dotnet/roslyn/issues/32662
- https://github.com/dotnet/roslyn/issues/37981

**Comment by GeeLaw** [comment]
It appears the scope of the [original comment in 85201](https://github.com/dotnet/roslyn/pull/85201#issuecomment-5589135787) is covered by #37981 but this issue contains more analysis --- notably clarification on pointer and byref fields. AFAIK currently Roslyn doesn't report IDE0044 on pointer/byref fields.

**Comment by CyrusNajmabadi** [comment]
Thanks for writing this up. I think the bulk of the concern here is already covered by https://github.com/dotnet/roslyn/issues/37981, so I’m fine with moving the general analysis there. It also seems reasonable to keep this issue open specifically for the pointer/byref cases, which are distinct and not covered as directly by that issue.

For context, when we originally designed and evaluated IDE0044’s behavior here, we ran the analyzer across Roslyn, dotnet/runtime, the Visual Studio codebase, other internal first and second party codebases, and more than 50 of the most widely used NuGet packages.

That analysis found only a low-single-digit number of cases that genuinely required mutable, unconstrained generic fields. By contrast, there were orders of magnitude more cases where making those fields readonly was correct and desirable. In practice, the existing heuristic was right more than 99.9% of the time. Taking the more conservative approach would have suppressed a useful suggestion across a very large number of valid cases to avoid a tiny number of exceptions, making the analyzer substantially less useful and more frustrating overall.

We recognize that the exceptional case is annoying when someone encounters it, particularly because changing the field to readonly can alter behavior rather than merely affect style. This tradeoff was considered carefully, though, using a broad corpus of real-world code. Based on that evidence, we believe the existing treatment of unconstrained generics should remain.

**Comment by hamarb123** [comment]
Fwiw, my defensive copies analyzer can catch if you accidentally write code like `myReadOnlyField.Init()` if you want to ensure you can catch it: https://github.com/hamarb123/hamarb123.Analyzers#defensive-copies-analysers-c-only.

If you mark the field as `readonly`, then you should get `HAM0001: Call to member 'Init' will execute on a defensive copy of 'foo'`, which would then indicate you should suppress the `IDE0044` suggestion if you have it set as error or warning or similar & not make it `readonly`, or you can suppress the `HAM0001` warning if you know it's wrong in your particular case (or configure it to some other severity in the `.editorconfig` iirc - I personally have it set as error via TreatWarningsAsError).
